net/mwan3/files/etc/hotplug.d/firewall/mwan3-restart

```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+if [ "$ACTION" = "restart" ]; then
```

Comment:
have you test this?
I try to /etc/init.d/firewall restart
but not ACTION=restart called
just somewhat `add` `remove` ACTION

Reply:
Good hint. I completely forgot that this is a patch set of mine that is not upstream.t was self-evident to me that this would work! 
https://github.com/lede-project/source/pull/1311
For some reason, which I can not restore in my mind, I closed the PR and no attention was paid to it anymore.
Maybe I should try this again.

It was probably due to the lede and openwrt remerge!

Reply:
@stangri As I see this your code does not work as well https://github.com/openwrt/packages/blob/master/net/vpnbypass/files/vpnbypass.hotplug? Or do I miss something.

Reply:
Sounds like the hotplug solution used in vpnbypass and vpn-policy-routing would not be working either.

If this is going to require changing other packages, it would be great to add triggers for this sort of thing in procd. Otherwise we can modify the firewall config using uci-defaults like @ptp52 suggests.

Reply:
I would prefer the hotplug.d, as it does not require any configuration changes. But that means that we have to get the change into fw3 first. As I said before, I tried that but I didn't finish it. 
Also, nobody has tried my patches, which are still applicable by the way.
In the source of the fw3 nothing has changed in this area so far.

@ptpt52 and @aaronjg Can you also test this [P/R](https://github.com/lede-project/source/pull/1311/) if it con work like this on your own builds?

Reply:
Yeah, I too wouldd prefer not to change the firewall config. But if we are going to change an upstream package, why not change procd so that one init script can have triggers on restart of other init scripts? This seems like the more general solution.

Reply:
procd is also an upstream package.
But that would also be a possibility for us to adapt the procd or /etc/rc.common
We would have to extend the procd so that it calls the scripts in /etc/hotplug.d/procd/* with the parameters SERVICE and ACTION when start/stop/restart/reload from a service is executed.

Reply:
`procd` already has facilities to handle triggers on config changes and on interface changes, so rather than creating a new hotplug directory, we could hook into these existing facilities.

We could add a `procd` trigger that would be fired on procd events, and then in `/etc/init.d/mwan3` `service_triggers()` add something like:

`procd_add_procd_trigger firewall restart|start reload_firewall_rules`

This would register the trigger with procd, and then when an event happens on the associated procd module, it would fire the trigger for mwan3.

Reply:
> @stangri As I see this your code does not work as well https://github.com/openwrt/packages/blob/master/net/vpnbypass/files/vpnbypass.hotplug? Or do I miss something.

Thanks for pinging me, I just tested this myself and I don't see the `/etc/hotplug.d/firewall/99-vpn-policy-routing` called at all, with any parameters on neither fw reload/restart. I may have used incorrect syntax for the `procd_add_procd_trigger` as well, because I don't see VPR being reloaded after firewall restart either. I haven't tested this part for a while, not sure when it broke.

Reply:
> I may have used incorrect syntax for the procd_add_procd_trigger as well, 

Just to be clear, in my previous comment, I was suggesting an extension to procd to handle this. No such feature exists yet as far as I know

Reply:
> > I may have used incorrect syntax for the procd_add_procd_trigger as well,
> 
> Just to be clear, in my previous comment, I was suggesting an extension to procd to handle this. No such feature exists yet as far as I know

I was under impression that things like `procd_add_procd_trigger firewall` in `service_triggers()` are supposed to work in current releases/snapshots. Are they not?

Reply:
I don’t see any such code for that kind of trigger:

https://github.com/openwrt/openwrt/blob/master/package/system/procd/files/procd.sh#L556

Reply:
Sorry, confused `procd_add_procd_trigger` and `procd_add_reload_trigger`, of which only the latter works. 

If the PROCD fix is merged, how likely that it will be backported to current releases?

Reply:
According to the documentation, the reload_trigger is only fired on a config change, so that is not what we want here.

Reply:
@aaronjg @stangri  I have added to your function the following code.
```
        blob_buf_init(&b, 0);
        blobmsg_add_string(&b, "service", s->name);
        trigger_event("service.restart", b.head);
```
I have update the staged [P/R](https://github.com/TDT-AG/openwrt/tree/pr/20201124-procd-restart) with my changes in my forked openwrt repository.
The following changes must also be done in mwan3 so the callback is execute on an firewall restart.

```
extra_command "test" "This is a test callback"
test() {
  logger "mwan3: Firewall reload execute test"
}

service_triggers() {
  procd_open_trigger
  procd_add_service_trigger "service.restart" "firewall" /etc/init.d/mwan3 test
  procd_add_config_trigger "config.change" "mwan3" /etc/init.d/mwan3 reload
  procd_close_trigger
}
```

And now I get an event which triggers my mwan3 callback on a firewall restart.
Everything is wrapped in the new restart method for the service ubus object.
I also get the notify.
Whatever I need it for.
I am not a ubus specialist so maybe someone could explain me for what this notify is for.